src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,53 +1,120 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 /**
  * <p>Реалізувати свій ArrayList який імплементує інтерфейс List. Дотриматися основних вимог щодо
  * реалізації ArrayList (default capacity, newCapacity...)</p>
  */
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private int nextEmptyPosition;
+    private T[] elements;
+
+    public ArrayList() {
+        this.nextEmptyPosition = 0;
+        this.elements = (T[]) new Object[DEFAULT_CAPACITY];
+    }
 
     @Override
     public void add(T value) {
-
+        checkAndIncreaseCapacity();
+        elements[nextEmptyPosition++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index < 0 || index > nextEmptyPosition) {
+            throw new ArrayIndexOutOfBoundsException();
+        }
+        checkAndIncreaseCapacity();
+        shiftToTheRightFrom(index);
+        nextEmptyPosition++;
+        elements[index] = value;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        while (list.size() > elements.length - nextEmptyPosition - 1) {
+            increaseCapacity();
+        }
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexInRange(index);
+        return elements[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndexInRange(index);
+        elements[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexInRange(index);
+        T removedElement = elements[index];
+        if (index != nextEmptyPosition) {
```

Comment:
try to test this without this check. Maybe it will work and you will just trim the empty space from the end of array

Reply:
yes, it works. it was redundant check